drivers/ipm/ipm_psoc6.c

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright (c) 2018, Cypress Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <device.h>
+#include <string.h>
+#include <ipm.h>
+#include <soc.h>
+#include "cy_sysint.h"
+#include "cy_ipc_drv.h"
+#include "ipm_psoc6.h"
+
+typedef enum {
+	PSOC6_IPM_SERVER = 0,
+	PSOC6_IPM_CLIENT = 1
+} psoc6_ipm_instance_role_t;
+
+struct psoc6_ipm_config_t {
+	u32_t channel;
+	s32_t int_base;
+	s32_t notify_chn;
+	s32_t release_chn;
+	psoc6_ipm_instance_role_t role;
+};
+
+struct psoc6_ipm_data {
+	ipm_callback_t callback;
+	void *callback_ctx;
+};
+
+static void psoc6_ipm_notify_isr(void *arg)
+{
+	struct device *dev = arg;
+	struct psoc6_ipm_data *data = dev->driver_data;
+	const struct psoc6_ipm_config_t *config = dev->config->config_info;
+	u32_t value;
+	u32_t interruptMasked;
+
+	/*
+	 * Check that there is really the IPC Notify interrupt,
+	 * because the same line can be used for the IPC Release interrupt.
+	 */
+	interruptMasked = Cy_IPC_Drv_ExtractAcquireMask(
+		Cy_IPC_Drv_GetInterruptStatusMasked(
+			Cy_IPC_Drv_GetIntrBaseAddr(config->notify_chn)));
+
+	if (interruptMasked == (1uL << config->channel)) {
+		Cy_IPC_Drv_ClearInterrupt(
+			Cy_IPC_Drv_GetIntrBaseAddr(config->notify_chn),
+				CY_IPC_NO_NOTIFICATION, interruptMasked);
+
+		if (Cy_IPC_Drv_ReadMsgWord(
+			Cy_IPC_Drv_GetIpcBaseAddress(config->channel), &value) ==
+			CY_IPC_DRV_SUCCESS) {
+			if (data->callback) {
+				data->callback(data->callback_ctx, 0, &value);
+			}
+		/* Release the REE IPC channel with the Release interrupt */
+			Cy_IPC_Drv_LockRelease(
+				Cy_IPC_Drv_GetIpcBaseAddress(config->channel),
+				(1uL << config->release_chn));
+		}
+	}
+}
+
+static void psoc6_ipm_release_isr(void *arg)
+{
+	struct device *dev = arg;
+	struct psoc6_ipm_data *data = dev->driver_data;
+	const struct psoc6_ipm_config_t *config = dev->config->config_info;
+	u32_t interruptMasked;
+
+	interruptMasked = Cy_IPC_Drv_ExtractReleaseMask(
+		Cy_IPC_Drv_GetInterruptStatusMasked(
+			Cy_IPC_Drv_GetIntrBaseAddr(config->release_chn)));
+
+	if ((interruptMasked & (1uL << config->channel)) != 0u) {
+		Cy_IPC_Drv_ClearInterrupt(
+			Cy_IPC_Drv_GetIntrBaseAddr(config->release_chn),
+			(1uL << config->channel), CY_IPC_NO_NOTIFICATION);
+
+		if (data->callback) {
+			data->callback(data->callback_ctx, 0, NULL);
+		}
+	}
+}
+
+static int psoc6_ipm_send(struct device *d, int wait, u32_t id,
+			const void *data, int size)
+{
+	const struct psoc6_ipm_config_t *config = d->config->config_info;
+	IPC_STRUCT_Type *ipcBase;
+	cy_en_ipcdrv_status_t ipcStatus;
+	u32_t data32;
+	int flags;
+
+	ARG_UNUSED(wait);
```

Comment:
Maybe should return -EINVAL if wait is non-zero.

Or is it possible to check whether the message has been consumed by the remote side? I am not familiar with this SoC, maybe Cy_IPC_Drv_GetLockStatus() could indicate that?

Reply:
Not sure about the wait parameter. The code here is made similarly to MCUX code.
In PSoC6 case there is Cy_IPC_Drv_LockAcquire() used inside the psoc6_ipm_send() function. In a case the channel is still not released, the psoc6_ipm_send() returns -EBUSY.
There are Cy_IPC_Drv_IsLockAcquired() and Cy_IPC_Drv_GetLockStatus() functions to check the IPC channel status also.

Reply:
The user application code will use the function ipm_send(). It is described [here](https://docs.zephyrproject.org/latest/api/io_interfaces.html):

> wait: If nonzero, busy-wait for remote to consume the message. The message is considered consumed once the remote interrupt handler finishes. If there is deferred processing on the remote side, or you would like to queue outgoing messages and wait on an event/semaphore, you can implement that in a high-level driver 

Return Value

> EBUSY: If the remote hasn’t yet read the last data sent. 

So the value EBUSY could be returned before sending the message. If the wait is nonzero, waiting is done after sending the message. If wait is set, ipm_send() blocks and a next call to ipm_send() then would not return EBUSY. You can have a look at function imx_mu_ipm_send() in ipm_imx.c in this [pull request](https://github.com/zephyrproject-rtos/zephyr/pull/8527/commits/feb9cc800179e88fcdf87495dc36d45ca59a8402).